runtime/win32.c

```diff
@@ -1272,3 +1271,32 @@ value caml_win32_xdg_defaults(void)
 
   CAMLreturn(result);
 }
+
+static DWORD (WINAPI *get_temp_path)(DWORD, LPWSTR);
+static INIT_ONCE get_temp_path_init_once = INIT_ONCE_STATIC_INIT;
+static BOOL CALLBACK get_temp_path_init_function(PINIT_ONCE InitOnce,
+                                                 PVOID Parameter,
+                                                 PVOID *lpContext)
+{
+  get_temp_path = (DWORD (WINAPI *)(DWORD, LPWSTR))
+    GetProcAddress(GetModuleHandle(L"KERNEL32.DLL"), "GetTempPath2W");
+  if (!get_temp_path)
+    get_temp_path = GetTempPath;
+  return TRUE;
+}
```

Comment:
How come it's being done this way? Isn't the point of using `InitOnceExecuteOnce` that you can use `lpContext` to get it as a local?
I've not tried it, but I was expecting something on the lines of:
```suggestion
static INIT_ONCE get_temp_path_init_once = INIT_ONCE_STATIC_INIT;
static BOOL CALLBACK get_temp_path_init_function(PINIT_ONCE InitOnce,
                                                 PVOID Parameter,
                                                 PVOID *lpContext)
{
  FARPROC pGetTempPath2W =
    GetProcAddress(GetModuleHandle(L"KERNEL32.DLL"), "GetTempPath2W");
  if (pGetTempPath2W)
    *lpContext = pGetTempPath2W;
  else
    *lpContext = GetTempPath;
  return TRUE;
}
```
and then below...

Reply:
I'd agree, but the documentation states:

> A parameter that receives data stored with the one-time initialization structure upon success. The low-order `INIT_ONCE_CTX_RESERVED_BITS` bits of the data are always zero. If `Context` points to a data structure, the data structure must be `DWORD`-aligned. `Context` must not be a code pointer on Arm32, because Arm32 code pointers always have the least significant bit set, see the [Arm32 ABI](https://learn.microsoft.com/en-us/cpp/build/overview-of-arm-abi-conventions?view=msvc-170#instruction-set) for details.

I understood "code pointers" as function pointer, is that wrong? I know we're not supporting Arm32 right now,  but that worries me a bit for the future.

Reply:
We're never going to support arm32 on Windows - apart from anything else, the support has gone.

Reply:
Raymond Chen presented some examples: [What does `INIT_ONCE_CTX_RESERVED_BITS` mean?](https://devblogs.microsoft.com/oldnewthing/20120420-00/?p=7813).

> `INIT_ONCE_CTX_RESERVED_BITS` value is the number of bits that the one-time initialization code uses after initialization is complete; the other bits are free for you to use yourself. The value of `INIT_ONCE_CTX_RESERVED_BITS` is 2, which means that you can store any value that’s a multiple of 4. If it’s a pointer, then the pointer must be `DWORD`-aligned or better.

[Overview of ARM32 ABI Conventions § Instruction Set](https://learn.microsoft.com/en-us/cpp/build/overview-of-arm-abi-conventions?view=msvc-170#instruction-set) reads:

> The instruction set for Windows on ARM is strictly limited to Thumb-2. […]
>
> A side-effect of this requirement is that all code pointers must have the low bit set. Then, when they're loaded and branched to via BLX or BX, the processor remains in Thumb mode. It doesn't try to execute the target code as 32-bit ARM instructions.

Is that something we might eventually care about?

Reply:
> We're never going to support arm32 on Windows - apart from anything else, the support has gone.

Noted, thanks! I'll use the `Context` parameter, then.

Reply:
I've applied your suggestion.